internal/controllers/ws_test.go

```diff
@@ -0,0 +1,191 @@
+			expectedResponses:      []string{},
+			forceNodeWStoFail:      true,
+			penalizeNodeNumOfCalls: 1,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			nodeRepoMock := mocks.NodeRepository{}
+			nodeRepoMock.On("GetActiveNodes", mock.Anything).Return(&test.nodeRepoGetActiveNodesReturn)
+			nodeRepoMock.On("UpdateNodeUsed", mock.Anything).Return()
+
+			recordRepoMock := mocks.RecordRepository{}
+			recordRepoMock.On("Save", mock.Anything).Return(nil)
+
+			actionsMockObject := new(actionMocks.Actions)
+			actionsMockObject.On(
+				"PenalizeNode", mock.MatchedBy(func(n models.Node) bool { return n.ID == "1" }), mock.Anything,
+			).Return()
+
+			apiController := NewApiController(false, repositories.Repos{
+				NodeRepo:   &nodeRepoMock,
+				RecordRepo: &recordRepoMock,
+			}, actionsMockObject)
+
+			// start test loadbalancer ws server
+			router := mm.NewRouter()
+			router.HandleFunc("/ws", apiController.WSHandler)
+			s := httptest.NewServer(router)
+			defer s.Close()
+
+			// start mock node ws server
+			ns := StartMockNodeWS(test.forceNodeWStoFail)
+			defer ns.Close()
+
+			// mock pooler
+			strPort := strings.Split(strings.Split(ns.URL, ":")[2], "/")[0]
+			nodePort, _ := strconv.Atoi(strPort)
+			poolerMock := tunnelMocks.Pooler{}
+			poolerMock.On("GetWSPort", mock.Anything).Return(nodePort, nil)
+			configuration.Config.PortPool = &poolerMock
+
+			// ws dial loadbalancer
+			u := "ws" + strings.TrimPrefix(s.URL, "http") + "/ws"
+			ws, _, err := websocket.DefaultDialer.Dial(u, nil)
+			assert.NoError(t, err)
+			defer ws.Close()
+
+			go func() {
+				for _, respons := range test.expectedResponses {
+					_, msg, err2 := ws.ReadMessage()
+					assert.NoError(t, err2)
+					assert.Equal(t, respons, string(msg))
+				}
+			}()
+
+			_ = ws.WriteMessage(1, []byte(test.requestType))
+			time.Sleep(1 * time.Second)
+			actionsMockObject.AssertNumberOfCalls(t, "PenalizeNode", test.penalizeNodeNumOfCalls)
+			nodeRepoMock.AssertNumberOfCalls(t, "UpdateNodeUsed", test.updateNodeUsedNumOfCalls)
+
+			// cleanup
+			configuration.Config.PortPool = nil
+		})
+	}
+}
+
+func StartMockNodeWS(shouldFail bool) *httptest.Server {
+	mockNodeWs := MockNodeWs{shouldFail: shouldFail}
+	nodeRouter := mm.NewRouter()
+	nodeRouter.HandleFunc("/", mockNodeWs.MockEchoWsHandler)
+	return httptest.NewServer(nodeRouter)
+}
+
+const (
+	SimpleRequest    = "request"
+	SubscribeRequest = "subscribe"
+	FailRequest      = "fail"
+)
+
+type MockNodeWs struct {
+	shouldFail bool
+}
+
+func (n *MockNodeWs) MockEchoWsHandler(w http.ResponseWriter, r *http.Request) {
+	if n.shouldFail {
+		http.Error(w, "Failed establishing connection", 500)
+		return
+	}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		http.Error(w, "Failed upgrading connection", 500)
+		return
+	}
+	for {
+		msgType, msg, err := conn.ReadMessage()
+		if err != nil {
+			return
+		}
+		fmt.Println(string(msg))
```

Comment:
I suppose this is for debugging

Reply:
Yeah, but no need for this to be left out in code. Removed it, but unfortunately now @mpetrun5  you need to approve once again.

Reply:
@mpetrun5 :dash: you are fast :dash:

Reply:
Already expected that